File: clients/dotnet/src/Varta.Client/Status.cs
namespace Varta;

/// <summary>
/// Health status carried in a Varta beat frame. The byte value is the
/// VLP v0.2 wire encoding (see <c>book/src/spec/vlp.md</c>).
/// </summary>
/// <remarks>
/// <see cref="Status"/> deliberately omits <c>Stall</c> (0x03) — that
/// variant is observer-synthesised and is rejected on the wire by
/// <see cref="Frame.Decode"/> as <see cref="DecodeErrorKind.StallOnWire"/>.
/// </remarks>
public enum Status : byte
{
    /// <summary>Agent is healthy.</summary>
    Ok = 0x00,

    /// <summary>Agent is operational but degraded.</summary>
    Degraded = 0x01,

    /// <summary>Agent is in a critical state.</summary>
    Critical = 0x02,
}

File: clients/java/lib/src/main/java/health/varta/Status.java
package health.varta;

/**
 * Agent-emitted health status carried on byte offset 3 of every VLP frame.
 *
 * <p>{@code STALL} is observer-synthesized and never appears on the wire;
 * the decoder rejects it as {@link DecodeErrorKind#STALL_ON_WIRE}.</p>
 */
public enum Status {
    /** Wire byte {@code 0x00}. */
    OK((byte) 0x00),
    /** Wire byte {@code 0x01}. */
    DEGRADED((byte) 0x01),
    /** Wire byte {@code 0x02}. */
    CRITICAL((byte) 0x02);

    private final byte wireByte;

    Status(byte wireByte) {
        this.wireByte = wireByte;
    }

    public byte wireByte() {
        return wireByte;
    }

    /** Wire byte 0x03, reserved for observer-synthesized STALL. Never agent-emitted. */
    public static final byte STALL_WIRE_BYTE = (byte) 0x03;

    public static Status fromWireByte(byte b) {
        return switch (b) {
            case 0x00 -> OK;
            case 0x01 -> DEGRADED;
            case 0x02 -> CRITICAL;
            default -> null;
        };
    }
}
